### PR TITLE
gcsfuse cache method implementation

### PR DESCRIPTION
**Pull request by thepenguinco**
Implement addReplace, remove and get methods to interact with the cache.

Run go test ./...

There should be no failures.

TODO: implement some basic unit tests later to verify correctness of the cache datastructure.

### CODE REVIEW

File: internal/contentcache/contentcache.go [hunk 1]
```diff
@@ -50,3 +50,38 @@ func New(tempDir string, mtimeClock timeutil.Clock) *ContentCache {
...
+		BucketName: cacheObjectKey.BucketName,
+		ObjectName: cacheObjectKey.ObjectName,
+		Generation: generation,
+	}
+	file, err := c.NewCacheFile(rc, metadata)
+	c.fileMap[*cacheObjectKey] = file
+	return file, err
+}
+
+// Retrieve temp file from the cache
```
Comment (lezh) [inline internal/contentcache/contentcache.go:69]: Let's follow [the golang guidance](https://go.dev/doc/effective_go#commentary) that recommends the function comment to start with the function name. 

>  If every [doc](https://go.dev/cmd/go/#hdr-Show_documentation_for_package_or_symbol) comment begins with the name of the item it describes, you can use the doc subcommand of the [go](https://go.dev/cmd/go/) tool and run the output through grep.

So, we can change it to
```go
// Get retrieves a temp file from the cache given the GCS object name and its generation.
```

File: internal/contentcache/contentcache.go [hunk 2]
```diff
@@ -50,3 +50,38 @@ func New(tempDir string, mtimeClock timeutil.Clock) *ContentCache {
 func (c *ContentCache) NewTempFile(rc io.ReadCloser) (gcsx.TempFile, error) {
 	return gcsx.NewTempFile(rc, c.tempDir, c.mtimeClock)
 }
+
+// Function to add or update existing cache file
```
Comment (lezh) [inline internal/contentcache/contentcache.go:54]: The cache is going to be used concurrently. Any idea how to ensure thread safety? If the caller is required to ensure it's used in a thread safe way, we should note that in the comment.

File: internal/contentcache/contentcache.go [hunk 3]
```diff
@@ -50,3 +50,38 @@ func New(tempDir string, mtimeClock timeutil.Clock) *ContentCache {
...
+func (c *ContentCache) AddOrReplace(cacheObjectKey *CacheObjectKey, generation int64, rc io.ReadCloser) (gcsx.TempFile, error) {
+	if _, exists := c.fileMap[*cacheObjectKey]; exists {
+		c.fileMap[*cacheObjectKey].Destroy()
+	}
+	metadata := &gcsx.TempFileObjectMetadata{
+		BucketName: cacheObjectKey.BucketName,
+		ObjectName: cacheObjectKey.ObjectName,
+		Generation: generation,
+	}
+	file, err := c.NewCacheFile(rc, metadata)
```
Comment (lezh) [inline internal/contentcache/contentcache.go:64]: If there is an error, should we proceed with the file?

File: internal/contentcache/contentcache.go [hunk 4]
```diff
@@ -50,3 +50,38 @@ func New(tempDir string, mtimeClock timeutil.Clock) *ContentCache {
 func (c *ContentCache) NewTempFile(rc io.ReadCloser) (gcsx.TempFile, error) {
 	return gcsx.NewTempFile(rc, c.tempDir, c.mtimeClock)
 }
+
+// Function to add or update existing cache file
+func (c *ContentCache) AddOrReplace(cacheObjectKey *CacheObjectKey, generation int64, rc io.ReadCloser) (gcsx.TempFile, error) {
+	if _, exists := c.fileMap[*cacheObjectKey]; exists {
```
Comment (lezh) [inline internal/contentcache/contentcache.go:56]: If we already extracted a value from the map, why do we do it again in the next line?

**Review by lezh — CHANGES_REQUESTED** [review]

File: internal/contentcache/contentcache.go (hunk 2 above)
Comment (thepenguinco) [reply internal/contentcache/contentcache.go:54]: Will add a comment, this in memory map is not concurrent safe.

File: internal/contentcache/contentcache.go (hunk 3 above)
Comment (thepenguinco) [reply internal/contentcache/contentcache.go:65]: We do not include it in our map and return err to the caller.

File: internal/contentcache/contentcache.go [hunk 5]
```diff
@@ -51,6 +52,38 @@ func (c *ContentCache) NewTempFile(rc io.ReadCloser) (gcsx.TempFile, error) {
 	return gcsx.NewTempFile(rc, c.tempDir, c.mtimeClock)
 }
 
+// Add or replace creates a new cache file or updates an existing cache file
```
Comment (lezh) [inline internal/contentcache/contentcache.go:55]:
```
// AddOrReplace creates a new cache file or updates an existing cache file
```

File: internal/contentcache/contentcache.go [hunk 6]
```diff
@@ -51,6 +52,38 @@ func (c *ContentCache) NewTempFile(rc io.ReadCloser) (gcsx.TempFile, error) {
...
+		cacheObject.Destroy()
+	}
+	metadata := &gcsx.TempFileObjectMetadata{
+		BucketName: cacheObjectKey.BucketName,
+		ObjectName: cacheObjectKey.ObjectName,
+		Generation: generation,
+	}
+	file, err := c.NewCacheFile(rc, metadata)
+	if err != nil {
+		return nil, err
```
Comment (lezh) [inline internal/contentcache/contentcache.go:67]: Use `fmt.Errorf` to explain where the error comes from

File: internal/contentcache/contentcache.go [hunk 7]
```diff
@@ -51,6 +52,38 @@ func (c *ContentCache) NewTempFile(rc io.ReadCloser) (gcsx.TempFile, error) {
...
+// Get retrieves a file from the cache given the GCS object name and bucket name
+func (c *ContentCache) Get(cacheObjectKey *CacheObjectKey) (gcsx.TempFile, bool) {
+	file, exists := c.fileMap[*cacheObjectKey]
+	return file, exists
+}
+
+// Remove removes and destroys the specfied cache file and metadata on disk
+func (c *ContentCache) Remove(cacheObjectKey *CacheObjectKey) {
+	if _, exists := c.fileMap[*cacheObjectKey]; exists {
+		c.fileMap[*cacheObjectKey].Destroy()
```
Comment (lezh) [inline internal/contentcache/contentcache.go:82]: This is mentioned in a previous comment still unresolved. Unnecessary map lookup.

**Review by lezh — CHANGES_REQUESTED** [review]

File: internal/contentcache/contentcache.go [hunk 8]
```diff
@@ -51,6 +53,38 @@ func (c *ContentCache) NewTempFile(rc io.ReadCloser) (gcsx.TempFile, error) {
...
+	file, exists := c.fileMap[*cacheObjectKey]
+	return file, exists
+}
+
+// Remove removes and destroys the specfied cache file and metadata on disk
+func (c *ContentCache) Remove(cacheObjectKey *CacheObjectKey) {
+	if cacheObject, exists := c.fileMap[*cacheObjectKey]; exists {
+		cacheObject.Destroy()
+	}
+	delete(c.fileMap, *cacheObjectKey)
```
Comment (lezh) [inline internal/contentcache/contentcache.go:85]: Why do we delete the entry even if the key does not exist?

**Review by lezh — CHANGES_REQUESTED** [review]

File: internal/contentcache/contentcache.go (hunk 8 above)
Comment (thepenguinco) [reply internal/contentcache/contentcache.go:85]: Delete should be a no-op if the key doesn't exist in the map, but I can move it inside.

Comment (lezh) [reply internal/contentcache/contentcache.go:85]: Please move it inside for better readability and less overhead.

**Pull request merged**